### PR TITLE
[mlir][nfc] Add tests for linalg.mmt4d

### PR DESCRIPTION
**Pull request by banach-space**
linalg.mmt4d was added a while back (https://reviews.llvm.org/D105244),
but there are virtually no tests in-tree. In the spirit of documenting
through test, this PR adds a few basic examples.

### CODE REVIEW

File: mlir/test/Dialect/Linalg/transform-op-mmt4d-to-fma.mlir [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    // Hoisting and LICM - not strictly required
+    %func_h = transform.structured.hoist_redundant_vector_transfers %func
+      : (!transform.op<"func.func">) -> !transform.op<"func.func">
+    %all_loops = transform.structured.match interface{LoopLikeInterface} in %func_h
+      : (!transform.op<"func.func">) -> !transform.any_op
+    transform.apply_licm to %all_loops : !transform.any_op
+    transform.loop.hoist_loop_invariant_subsets %all_loops : !transform.any_op
+
+    // Simplify the 6-dim vector.contract into a 3-dim matmul-like
+    // vector.contract with the following split splitn parallel and reduction
```
Comment (c-rhodes) [inline mlir/test/Dialect/Linalg/transform-op-mmt4d-to-fma.mlir:54]:
```suggestion
    // vector.contract with the following split of parallel and reduction
```
?

File: mlir/test/Dialect/Linalg/transform-op-mmt4d-to-fma.mlir [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    // dims:
+    //    * parallel, parallel, reduction
+    transform.apply_patterns to %func_h {
+      transform.apply_patterns.vector.reduction_to_contract
+      transform.apply_patterns.vector.cast_away_vector_leading_one_dim
+      transform.apply_patterns.canonicalization
+    } : !transform.op<"func.func">
+
+    // Step 4: Lower vector.contract to vector.fma
```
Comment (c-rhodes) [inline mlir/test/Dialect/Linalg/transform-op-mmt4d-to-fma.mlir:63]:
```suggestion
    // Step 4: Lower vector.contract to vector.fma via vector.outerproduct
```

**Review by c-rhodes — COMMENTED** [review]
thanks Andrzej, I'm not very familiar with this op so would be good to get eyes from someone who is, I've left a few comments anyway. I would also find a small integration test helpful, particularly if it could be compared with an existing linalg.matmul.

**Review by bjacob — APPROVED** [review]
Thanks for doing this, @banach-space! Sorry about the omission.

**Comment by banach-space** [comment]
> I would also find a small integration test helpful, particularly if it could be compared with an existing linalg.matmul.

That's what I plan next, but just run out of spare cycles when working on this :) Reverse engineering that lowering pipeline took quite some time.

File: mlir/test/Dialect/Linalg/transform-op-mmt4d-to-fma.mlir (hunk 1 above)
Comment (banach-space) [reply mlir/test/Dialect/Linalg/transform-op-mmt4d-to-fma.mlir]: "into" :)

**Review by c-rhodes — APPROVED** [review]
LGTM cheers

**Comment by banach-space** [comment]
> I would also find a small integration test helpful

https://github.com/llvm/llvm-project/pull/81790

**Pull request merged**